bin/external_lib_deps.ml

```diff
@@ -1,28 +1,217 @@
+  List.fold_left
+    ~f:(fun acc_map (dir, libs) ->
+      match Dir_map.find acc_map dir with
+      | None -> Dir_map.add_exn acc_map dir (libs_to_lib_map libs)
+      | Some libs_map ->
+        Dir_map.set acc_map dir
+          (Lib_name.Map.union libs_map (libs_to_lib_map libs) ~f:(fun _ k1 k2 ->
+               Some (Kind.merge k1 k2))))
+    ~init:Dir_map.empty libs_dir
+
+let libs_sort libs =
+  List.sort_uniq
+    ~compare:(fun (name1, _) (name2, _) ->
+      String.compare (Lib_name.to_string name1) (Lib_name.to_string name2))
+    libs
+
+let run ~context_name ~external_resolved_libs ~by_dir ~sexp ~only_missing =
+  if only_missing then (
+    if by_dir || sexp then
+      User_error.raise
+        [ Pp.textf "--only-missing cannot be used with --by_dir or --sexp" ];
+    let missing =
+      external_resolved_libs |> List.map ~f:snd |> List.concat
+      |> List.filter_map ~f:(fun ((result, kind_db), kind, name) ->
+             match (result, kind_db) with
+             | None, Lib.Kind_db.Installed_libs -> Some (name, kind)
+             | _ -> None)
+      |> libs_sort
+    in
+    if List.for_all ~f:(fun (_, kind) -> kind = Kind.Optional) missing then
+      User_message.prerr
+        (User_error.make
+           [ Pp.textf "The following libraries are missing in the %s context:"
+               context_name
+           ; pp_external_libs missing
+           ])
+    else
+      User_message.prerr
+        (User_error.make
+           [ Pp.textf "The following libraries are missing in the %s context:"
+               context_name
+           ; pp_external_libs missing
+           ]
+           ~hints:
+             [ Dune_engine.Utils.pp_command_hint
+               @@ (missing
+                  |> List.filter_map ~f:(fun (name, kind) ->
+                         match (kind : Kind.t) with
+                         | Optional -> None
+                         | Required -> Some (Lib_name.package_name name))
+                  |> List.map ~f:Package.Name.to_string
+                  |> String.concat ~sep:" ")
+             ]))
+  else if sexp then (
+    if not by_dir then User_error.raise [ Pp.textf "--sexp requires --by-dir" ];
+    let sexp =
+      external_resolved_libs
+      |> List.map ~f:(fun (dir, libs) ->
+             ( dir
+             , libs
+               |> List.filter_map ~f:(fun ((result, kind_db), kind, name) ->
+                      match (result, kind_db) with
+                      | _, Lib.Kind_db.Installed_libs -> Some (name, kind)
+                      | _ -> None) ))
+      |> libs_dir_to_map
+      |> Dir_map.filter ~f:(fun m -> not (Lib_name.Map.is_empty m))
+      |> Dir_map.to_dyn (fun libs -> Lib_name.Map.to_dyn Kind.to_dyn libs)
+      |> Sexp.of_dyn
+    in
+    Format.printf "%a@." Pp.to_fmt
+      (Sexp.pp (Sexp.List [ Sexp.Atom context_name; sexp ])))
+  else (
+    if by_dir then
+      User_error.raise [ Pp.textf "--by-dir cannot be used without --sexp" ];
```

Comment:
> I wonder if this should be a `describe` target rather than a top-level command.

I don't get what you mean.

Reply:
I was trying keep the old interface, there's no by-dir plain output. I asked myself if the `by-dir` is necessary because only `sexp` could be ok, I think.